src/pool/spawn.rs

```diff
@@ -1,56 +1,276 @@
+/// to avoid future breaks.
+trait AssertSync: Sync {}
+impl<T: Send> AssertSync for Remote<T> {}
+trait AssertSend: Send {}
+impl<T: Send> AssertSend for Remote<T> {}
+
+/// Spawns tasks to the associated thread pool.
+///
+/// It's different from `Remote` because it submits tasks to the local queue
+/// instead of global queue, so new tasks can take advantage of cache
+/// coherence.
+pub struct Local<T> {
+    id: usize,
+    local_queue: LocalQueue<T>,
+    core: Arc<QueueCore<T>>,
+}
+
+impl<T: TaskCell + Send> Local<T> {
+    pub(crate) fn new(id: usize, local_queue: LocalQueue<T>, core: Arc<QueueCore<T>>) -> Local<T> {
+        Local {
+            id,
+            local_queue,
+            core,
+        }
+    }
+
+    /// Spawns a task to the local queue.
+    pub fn spawn(&mut self, task: T) {
+        self.local_queue.push(task);
+    }
+
+    /// Spawns a task to the remote queue.
+    pub fn spawn_remote(&self, task: T) {
+        self.core.push(self.id, task);
+    }
+
+    /// Gets a remote so that tasks can be spawned from other threads.
+    pub fn remote(&self) -> Remote<T> {
+        Remote {
+            core: self.core.clone(),
+        }
+    }
+
+    pub(crate) fn core(&self) -> &Arc<QueueCore<T>> {
+        &self.core
+    }
+
+    pub(crate) fn pop(&mut self) -> Option<Pop<T>> {
+        self.local_queue.pop()
+    }
+
+    /// Pops a task from the queue.
+    ///
+    /// If there is no tasks at the moment, it will go to sleep until woken
+    /// up by other threads.
+    pub(crate) fn pop_or_sleep(&mut self) -> Option<Pop<T>> {
+        let address = &*self.core as *const QueueCore<T> as usize;
+        let mut task = None;
+        let mut timeout = Some(Instant::now() + self.core.config.max_idle_time);
+        let id = self.id;
+        loop {
+            let res = unsafe {
+                parking_lot_core::park(
+                    address,
+                    || {
+                        if timeout.is_some() && !self.core.mark_sleep() || self.core.is_shutdown() {
+                            return false;
+                        }
+                        task = self.local_queue.pop();
+                        task.is_none()
+                    },
+                    || {},
+                    |_, _| {},
+                    ParkToken(id),
+                    timeout,
+                )
+            };
+            return match res {
+                ParkResult::Unparked(_) | ParkResult::Invalid => {
+                    self.core.mark_woken();
+                    task
+                }
+                ParkResult::TimedOut => {
+                    timeout = None;
+                    continue;
+                }
+            };
         }
     }
 }
+
+/// Building remotes and locals from the given queue and configuration.
+///
+/// This is only for tests purpose so that a thread pool doesn't have to be
+/// spawned to test a Runner.
+#[doc(hidden)]
```

Comment:
It seems a bit rough to just hide it from docs. How about placing it in a `test_util` mod?

Reply:
Actually I think an ideal unit test of a runner should not depend on `QueueCore`, which is another rather complex component.

Reply:
> ...an ideal unit test of a runner ...

Due to the removal of `Spawn` trait, there has to be a way to create a `Local` type.

> ... is another rather complex component.

Indeed. So the function is provided to hide the complexity.